backend/src/defence.ts

```diff
@@ -207,117 +211,149 @@ function transformMessage(
 
 // detects triggered defences in original message and blocks the message if necessary
 async function detectTriggeredDefences(message: string, defences: Defence[]) {
-	// keep track of any triggered defences
-	const defenceReport: ChatDefenceReport = {
-		blockedReason: null,
-		isBlocked: false,
-		alertedDefences: [],
-		triggeredDefences: [],
-	};
+	const singleDefenceReports = [
+		detectCharacterLimit(message, defences),
+		detectFilterUserInput(message, defences),
+		detectXmlTagging(message, defences),
+		await detectEvaluationLLM(message, defences),
+	];
+
+	return combineDefenceReports(singleDefenceReports);
+}
 
-	// the following methods will add triggered defences to the defenceReport
-	detectCharacterLimit(defenceReport, message, defences);
-	detectFilterUserInput(defenceReport, message, defences);
-	detectXmlTagging(defenceReport, message, defences);
-	await detectEvaluationLLM(defenceReport, message, defences);
+function combineDefenceReports(
+	defenceReports: SingleDefenceReport[]
+): ChatDefenceReport {
+	const isBlocked = defenceReports.some((report) => report.blockedReason);
+	const blockedReason = isBlocked
+		? defenceReports
+				.map((report) => report.blockedReason)
+				.filter((reason) => reason !== null)
+				.join('\n')
```

Comment:
Not something you need to change, but just thought I'd point out it's _nanofractionally_ more efficient to filter and then map, rather than map and then filter.